rfcs/0000-limitador-multithread-inmemory.md

```diff
@@ -0,0 +1,241 @@
+have been denied. This scenario is less likely to happen, but it's still possible. In this case, the _undershoot_ would
+be the difference between the expected value of the counter and the value that the counter has in the storage, when the
+stored value is **lower** than the expected one. Usually, this would happen when trying to revert a wrongly updated counter
+twice in a row, for example, when trying to revert the _overshoot_ from the previous example.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Enhancing the Limitador service to process requests in parallel and being capable of operating in the previously described
+modes, entails considering various implications.
+
+Firstly, ensuring consistency in storing counter values across multiple threads is paramount to maintaining the integrity
+of rate limiting operations. This involves implementing robust concurrency control mechanisms to prevent race conditions
+and data corruption when accessing and updating shared counter data.
+
+Additionally, choosing the data structure that will store the counter values must prioritize efficiency and thread safety
+to handle concurrent access effectively.
+
+Finally, considering that initially we will implement the setup of the mode at the service initialization time, balancing
+the need for strict adherence to defined limits with the desire to maximize throughput presents a trade-off that we could
+give the user the ability to manage.
+
+
+## Implementation guidelines
+
+### Data structures
+
+Limitador already possess a data structure to store the limit counters that also stores the expiration time of the counter,
+the type is named `AtomicExpiringCounter`. This data structure is a thread-safe counter that can be incremented and
+decremented atomically, and it also has a methods to check the value at a certain time and update the counter.
+
+The collection of these counters should be consistent across threads and also being able to quickly sort and retrieve the
+counters associated to a certain namespace and limits definitions. Currently, the data structure can't provide this
+much desired functionality, so we need to introduce a new data structure that can provide this.
+
+### Request handling and concurrency control
+
+When a request comes in, Limitador needs to determine the appropriate Counter object(s) based on the request's namespace
+and limit definitions. Taking into account the collection will be in ascending order by expiration time, iterating over the
+counter objects associated with those premises and performing the following steps:
+
+1. Retrieve the Counter object from the collection.
+2. Compare the current timestamp with the expiration time of the counter to determine if the request falls within the
+sliding window
+3. If the request falls within the window, increment the hit count of the Counter.
```

Comment:
shouldn't the increment happen only when it has been verified that no counters has exceeded the associated limit?